sdk/typescript/introspector/scanner/typescript_module/ast.ts

```diff
@@ -0,0 +1,368 @@
+      return type.slice("Array<".length, -">".length)
+    }
+
+    return type
+  }
+
+  public stringTypeToUnwrappedType(type: string): string {
+    type = this.unwrapTypeStringFromPromise(type)
+
+    // If they are difference, that means we upwrapped the array.
+    const extractedTypeFromArray = this.unwrapTypeStringFromArray(type)
+    if (extractedTypeFromArray !== type) {
+      return this.stringTypeToUnwrappedType(extractedTypeFromArray)
+    }
+
+    return type
+  }
+
+  public typeToStringType(type: ts.Type): string {
+    const stringType = this.checker.typeToString(type)
+
+    return this.stringTypeToUnwrappedType(stringType)
+  }
+
+  public tsTypeToTypeDef(
+    node: ts.Node,
+    type: ts.Type,
+  ): TypeDef<TypeDefKind> | undefined {
+    if (type.flags & ts.TypeFlags.String)
+      return { kind: TypeDefKind.StringKind }
+    if (type.flags & ts.TypeFlags.Number)
+      return { kind: TypeDefKind.IntegerKind }
+    if (type.flags & ts.TypeFlags.Boolean)
+      return { kind: TypeDefKind.BooleanKind }
+    if (type.flags & ts.TypeFlags.Void) return { kind: TypeDefKind.VoidKind }
+
+    // If a type has a flag Object, is can basically be anything.
+    // We firstly wants to see if it's a promise or an array so we can unwrap the
+    // actual type.
+    if (type.flags & ts.TypeFlags.Object) {
+      const objectType = type as ts.ObjectType
+
+      // If it's a reference, that means it's a generic type like
+      // `Promise<T>` or `number[]` or `Array<T>`.
+      if (objectType.objectFlags & ts.ObjectFlags.Reference) {
+        const typeArguments = this.checker.getTypeArguments(
+          type as ts.TypeReference,
+        )
+
+        switch (typeArguments.length) {
+          case 0:
+            // Might change after to support more complex type
+            break
+          case 1: {
+            const typeArgument = typeArguments[0]
+
+            if (type.symbol.getName() === "Promise") {
+              return this.tsTypeToTypeDef(node, typeArgument)
+            }
+
+            if (type.symbol.getName() === "Array") {
+              return {
+                kind: TypeDefKind.ListKind,
+                typeDef: this.tsTypeToTypeDef(node, typeArgument),
+              }
+            }
+
+            return undefined
+          }
+          default: {
+            throw new IntrospectionError(
+              `could not resolve type ${type.symbol.getName()} at ${AST.getNodePosition(node)}, dagger does not support generics with argument yet.`,
+            )
+          }
+        }
+      }
+    }
+  }
+
+  private resolveParameterDefaultValueTypeReference(
+    expression: ts.Expression,
+    value: any,
+  ): any {
+    const type = typeof value
+
+    switch (type) {
+      case "string":
+      case "number":
+      case "bigint":
+      case "boolean":
+      case "object":
+        // Value will be jsonified on registration so it must be a string.
+        return `${value}`
```

Comment:
I don't understand this cast. Why not jsonify it?

Reply:
Because it's `jsonify` during the registration so we don't need to handle it here.

Reply:
Yes, but why the cast to string?

Reply:
Yeah you're right, I could leave it as plain value

Reply:
What do you mean? That doesn't make sense. Those are valid JSON primitives, so if you cast an integer to a string and then convert to JSON, you'll get a JSON string, not a JSON integer.

<img width="256" alt="Screenshot 2024-11-18 at 23 22 56" src="https://github.com/user-attachments/assets/b68125fe-e59c-4441-bbbf-26de69f04d8a">

Reply:
I can try to not stringify it and see what happens then

Reply:
[f805f57](https://github.com/dagger/dagger/pull/8824/commits/f805f575ead92372c4d457a82980c5259a1265bf) I'll see if the CI does well with that